File: src/Headings.hs
module Headings
  ( Heading(..)
  , readHeading
  ) where

data Heading
  = North
  | East
  | South
  | West
  deriving (Eq, Ord, Read)

instance Show Heading where
  show North = "N"
  show East  = "E"
  show South = "S"
  show West  = "W"

readHeading :: Char -> Maybe Heading
readHeading 'N' = Just North
readHeading 'E' = Just East
readHeading 'S' = Just South
readHeading 'W' = Just West
readHeading _   = Nothing

File: scala/src/main/scala/Headings.scala
package com.rumblesan.marsrover

object Headings {

  sealed trait Heading
  case object North extends Heading
  case object East extends Heading
  case object South extends Heading
  case object West extends Heading

  def parse(c: Char): Option[Heading] = {
    c match {
      case 'N' => Some(North)
      case 'E' => Some(North)
      case 'S' => Some(North)
      case 'W' => Some(North)
      case _ => None
    }
  }

}
